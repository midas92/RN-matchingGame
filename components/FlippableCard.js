import React, { useState, useEffect, useLayoutEffect, useRef } from 'react';
import { Animated, StyleSheet, Text, View, TouchableOpacity } from 'react-native';

const FlippableCard = ({
  value,
  onTilePressed,
  totalColumns,
  cardStyle,
  flipped,
  mode
}) => {

  const [animatedValue, setAnimatedValue] = useState(new Animated.Value(0));
  const [frontInterpolate, setFrontInterpolate] = useState("0deg");
  const [backInterpolate, setBackInterpolate] = useState("0deg");
  const [flipValue, setFlipValue] = useState(0);

  useEffect(() => {
    animatedValue.addListener(({ value }) => {
      setFlipValue(value)
    })

    setBackInterpolate(
      animatedValue.interpolate({
        inputRange: [0, 180],
        outputRange: ['0deg', '180deg']
      })
    )
    setFrontInterpolate(
      animatedValue.interpolate({
        inputRange: [0, 180],
        outputRange: ['180deg', '360deg']
      })
    )
  }, []);

  const firstUpdate = useRef(true);
  useLayoutEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false;
      return;
    }

    flipCard()
  }, [flipped]);

  const flipCard = () => {
    if( flipValue <= 90 ){
      Animated.timing(animatedValue, {
        toValue: 180,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start();
    } else {
      Animated.timing(animatedValue, {
        toValue: 0,
        friction: 8,
        tension: 10,
        useNativeDriver: true
      }).start();
    }
  }

  const frontAnimatedStyle = {
    transform: [
      { rotateX: frontInterpolate }
    ]
  }

  const backAnimatedStyle = {
    transform: [
      { rotateX: backInterpolate }
    ]
  }

  return(
    <View>
      <Animated.View style={[cardStyle, styles.flipCardBackVisibility, styles.flipCardBack, backAnimatedStyle]}>
        { mode === 'Dev' && <Text style={{fontSize: 125 / totalColumns }}>{value}</Text> }
      </Animated.View>
      <Animated.View style={[cardStyle, styles.flipCardBackVisibility, styles.flipCardFront, frontAnimatedStyle]}>
        <TouchableOpacity
          testID={'flipCardId'}
          style={[cardStyle]}
          onPress={() => onTilePressed()}
          disabled={flipped}
          activeOpacity={1.0}
        >
          <Text style={{fontSize: 125 / totalColumns }}>{value}</Text>
        </TouchableOpacity>
      </Animated.View>
    </View>
  );
}

const styles = StyleSheet.create({
  flipCardBackVisibility: {
    backfaceVisibility: 'hidden'
  },
  flipCardBack: {
    backgroundColor: 'gray',
  },
  flipCardFront: {
    position: 'absolute',
    backgroundColor: '#34a8eb',
  }
});

export default FlippableCard;
