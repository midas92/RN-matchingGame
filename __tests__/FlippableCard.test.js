import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import FlippableCard from '../components/FlippableCard';

const onTilePressed = jest.fn();

describe('<FlippableCard/>', () => {
  let result;
  beforeEach(() => result = render(
    <FlippableCard
      value={1}
      onTilePressed={onTilePressed}
      totalColumns={3}
      cardStyle={{}}
      flipped={false}
      mode={'normal'}
    />
    ));

  it('should match snapshot', () => expect(result.toJSON()).toMatchSnapshot());
  it('should render the FlippableCard component', () => expect(result).toBeTruthy());

  it('flips the card when pressed', () => {
    const flipCardPress = result.getByTestId('flipCardId');
    fireEvent(flipCardPress, 'press');
    expect(onTilePressed).toHaveBeenCalled;
  });
});
