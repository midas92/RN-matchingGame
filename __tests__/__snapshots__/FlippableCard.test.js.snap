// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlippableCard/> should match snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "backgroundColor": "gray",
        "transform": Array [
          Object {
            "rotateX": "0deg",
          },
        ],
      }
    }
  />
  <View
    style={
      Object {
        "backfaceVisibility": "hidden",
        "backgroundColor": "#34a8eb",
        "position": "absolute",
        "transform": Array [
          Object {
            "rotateX": "180deg",
          },
        ],
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="flipCardId"
    >
      <Text
        style={
          Object {
            "fontSize": 41.666666666666664,
          }
        }
      >
        1
      </Text>
    </View>
  </View>
</View>
`;
